content/blog/2020-12-19-LLVM-IR-to-assembly.md

```diff
@@ -0,0 +1,77 @@
+This begs the question of do we actually need the number of registers offered by the modern ISA? Or can we go back to the traditional AX, BX, CX, DX, SI, DI, BP, SP registers?  Although at the current landscape in O-3 processors the incorporation of registers are relative cheap, such relationships may not hold true in other alternative processing methods. (PIM, GPGPU, DNA computing, quantum computing, et cetera)  Therefore such dimensions are worth a closer inspection, and that we as researcher should not take the cheap overhead for registers in modern O3 processors for granted. 
+This serves as the motivation of this project, in which I will modify LLVM assembly generation routine to generate code with less registers than what is offered by the traditional LLVM compiler.  Furthermore, an informal comparison on the performance tradeoff based on the number of architectural registers is done.  
+
+# What did I do?
+Although initially the project was to target the RISC-V ISA, but the current RISC-V Vector extensions currently still experimental and has an [outstanding pull request in Phabricator](https://reviews.llvm.org/rG5baef6353e8819e443d327f84edc9f2d1c8c0c9e) as of writing this blog post.  The lack of support in both LLVM side and the experimental status on the vector extension makes it hard for someone new to LLVM to effectly modify the praser in a resonable time, therefore I have decide to folow Adrian Sampson's advice in switching to x86 AVX and using that to benchmark. The plus side is that it is more mature and widely supported in the developer mailing list and the community.  However in retrospect it is something that I deeply regret as x86-64 is significently more complex.  This project is largely done by modifying the LLVM x86-64 target con figurations to create new target machine with different number of architectural registers.  Although X86-64 has better support across the LLVM stack, modifying the x86 machine target register configurations is no easy task and is significantly harder due to the number of aliasing it has across the different registers.   Most of the work that I have done on this project is modifying the LLVM assembly code generation to ulitizing different number of registers.  Then by executing the complied output from gcc, llvm and modified llvm.  I hope to unearth the relationship of architectural registers and performance.  To compare the results across multiple register types, I have done microbenchmark using Matrix Multiplication with three different x86-64 implementation (scalar, SSE, AVX-128bit, AVX-256bit) and benchmarking with FFT computations.  
```

Comment:
"Then by executing the complied output from gcc, llvm and modified llvm." is a sentence fragment.